csvcubed/tests/unit/cli/build/test_build_with_config.py

```diff
@@ -0,0 +1,626 @@
+    UndefinedAttributeValueUrisError, UndefinedMeasureUrisError, ConflictingUriSafeValuesError, ReservedUriValueError
+from csvcubed.models.cube.qb.validationerrors import (
+    NoObservedValuesColumnDefinedError,
+    NoUnitsDefinedError,
+    NoMeasuresDefinedError,
+    MoreThanOneUnitsColumnError,
+    MoreThanOneMeasureColumnError,
+    MoreThanOneObservationsColumnError
+)
+from tests.unit.test_baseunit import get_test_cases_dir
+
+PROJECT_ROOT = Path(Path(__file__).parent, "..", "..", "..", "..").resolve()
+TEST_CASE_DIR = Path(get_test_cases_dir().absolute(), 'cli', 'build')
+SCHEMA_PATH_FILE = Path(PROJECT_ROOT, "csvcubed", "schema", "cube-config-schema.json")
+
+
+def test_01_01_build_ok():
+    config = Path(TEST_CASE_DIR, "cube_data_config_ok.json")
+    output = Path("./out")
+    csv = Path(TEST_CASE_DIR, "cube_data_config_ok.csv")
+    cube, validation_errors = cli_build(
+        config_path=config,
+        output_directory=output,
+        csv_path=csv,
+        fail_when_validation_error_occurs=False,
+        validation_errors_file_out=Path("validation_errors.json"),
+    )
+    assert isinstance(cube, Cube)
+    assert isinstance(validation_errors, list)
+    assert len(validation_errors) == 0
+
+
+def test_01_02_build_ok_fail_on_error():
+    config = Path(TEST_CASE_DIR, "cube_data_config_ok.json")
+    output = Path("./out")
+    csv = Path(TEST_CASE_DIR, "cube_data_config_ok.csv")
+    cube, validation_errors = cli_build(
+        config_path=config,
+        output_directory=output,
+        csv_path=csv,
+        fail_when_validation_error_occurs=True,
+        validation_errors_file_out=Path("validation_errors.json"),
+    )
+    assert isinstance(cube, Cube)
+    assert isinstance(validation_errors, list)
+    assert len(validation_errors) == 0
+
+
+def test_01_03_build_exit_on_fail():
+    """
+    Tests that the code exits on failure rather than returning incomplete cube with errors
+    """
+    config = Path(TEST_CASE_DIR, "cube_data_test_02_01.json")
+    output = Path("./out")
+    csv = Path(TEST_CASE_DIR, "cube_data_test_02_01.csv")
+    with pytest.raises(SystemExit) as err:
+        cube, validation_errors = cli_build(
+            config_path=config,
+            output_directory=output,
+            csv_path=csv,
+            fail_when_validation_error_occurs=True,
+            validation_errors_file_out=Path("validation_errors.json"),
+        )
+    assert err.value.code == 1
+
+
+def test_02_01_val_errors_no_observation():
+    """
+    Test for:-
+    NoObservedValuesColumnDefinedError
+    """
+    config = Path(TEST_CASE_DIR, "cube_data_test_02_01.json")
+    output = Path("./out")
+    csv = Path(TEST_CASE_DIR, "cube_data_test_02_01.csv")
+    cube, validation_errors = cli_build(
+        config_path=config,
+        output_directory=output,
+        csv_path=csv,
+        fail_when_validation_error_occurs=False,
```

Comment:
I think this is the default.

Reply:
When writing unit tests I normally spell out all params, even defaulted ones, for clarity.
I've now removed these.